impl/maven-core/src/main/java/org/apache/maven/internal/transformation/impl/DefaultConsumerPomBuilder.java

```diff
@@ -193,72 +202,72 @@ private ModelBuilderResult buildModel(RepositorySystemSession session, Path src)
-                warnNotDowngraded(project);
-            }
-            model = model.withModelVersion(modelVersion);
-        } else {
-            Model.Builder builder = prune(
-                            Model.newBuilder(model, true)
-                                    .preserveModelVersion(false)
-                                    .root(false)
-                                    .parent(null)
-                                    .build(null),
-                            model)
-                    .mailingLists(null)
-                    .issueManagement(null)
-                    .scm(
-                            model.getScm() != null
-                                    ? Scm.newBuilder(model.getScm(), true)
-                                            .childScmConnectionInheritAppendPath(null)
-                                            .childScmUrlInheritAppendPath(null)
-                                            .childScmDeveloperConnectionInheritAppendPath(null)
-                                            .build()
-                                    : null);
-            builder.profiles(prune(model.getProfiles()));
-
-            model = builder.build();
+        Model.Builder builder = prune(
+                        Model.newBuilder(model, true)
+                                .preserveModelVersion(false)
+                                .root(false)
+                                .parent(null)
+                                .build(null),
+                        model)
+                .mailingLists(null)
+                .issueManagement(null)
+                .scm(
+                        model.getScm() != null
+                                ? Scm.newBuilder(model.getScm(), true)
+                                        .childScmConnectionInheritAppendPath(null)
+                                        .childScmUrlInheritAppendPath(null)
+                                        .childScmDeveloperConnectionInheritAppendPath(null)
+                                        .build()
+                                : null);
+        builder.profiles(prune(model.getProfiles()));
+
+        model = builder.build();
+        String modelVersion = new MavenModelVersion().getModelVersion(model);
+        if (!ModelBuilder.MODEL_VERSION_4_0_0.equals(modelVersion) && !preserveModelVersion) {
+            warnNotDowngraded(project);
+        }
+        model = model.withModelVersion(modelVersion);
+        return model;
+    }
+
+    static Model transformBom(Model model, MavenProject project) {
+        boolean preserveModelVersion = model.isPreserveModelVersion();
+
+        Model.Builder builder = prune(
+                Model.newBuilder(model, true)
+                        .preserveModelVersion(false)
+                        .root(false)
+                        .parent(null)
+                        // .dependencyManagement(dependencyManagement.withDependencies(dependencies))
```

Comment:
why was this added at all? Was it envisioned that BOM may contain only reactor deps? As then we may want to enforce that

Reply:
See https://issues.apache.org/jira/browse/MNG-7920

Reply:
The problem is that the consumer POM for the BOM was originally created like a POM, i.e. from the raw model, and it was missing some versions for managed dependencies.  The fix was overwriting the dependency version with the model version.  This is now correctly implemented by using the effective model instead of the raw model.